File: AEDs/Algoritmos de Ordenacao/Java/Counting.java
public class Counting extends Geracao 
{
    public Counting ( )
    {
        super ( );
    }

    public Counting ( int tamanho )
    {
        super( tamanho );
    }

    public int getLargest ( ) 
    {
        int result = arranjo[0];
        for( int i = 0; i < tamanho; i++ ) 
        {
            if( result < arranjo[i] ) {
                result = arranjo[i];
            }
        }
        return ( result );	
    } // end getLargest ( )

    @Override
    public void sort( Log log  ) 
    {
        //arranjo para contar o numero de ocorrencias de cada elemento
        int[] count = new int[getLargest( ) + 1];
        int[] ordenado = new int[tamanho];
        log.incrementarSubarranjos(2);

        //Inicializar cada posicao do arranjo de contagem 
        for( int i = 0; i < count.length; count[i] = 0, i++ );
        
        //Agora, o count[i] contem o numero de elemento iguais a i
        for( int i = 0; i < tamanho; 
        count[arranjo[i]]++, log.incrementarMovimentacoes(), i++ );

        //Agora, o count[i] contem o numero de elemento menores ou iguais a i
        for( int i = 1; i < count.length; 
            count[i] += count[i-1], log.incrementarMovimentacoes(), i++ );

        //Ordenando
        for( int i = tamanho-1; i >= 0; 
            ordenado[count[arranjo[i]]-1] = arranjo[i], count[arranjo[i]]--, incrementarMovimentacoes(), i-- );

        //Copiando para o arranjo original
        for( int i = 0; i < tamanho; 
            arranjo[i] = ordenado[i], incrementarMovimentacoes(), i++ );
    } // end sort ( )

    public static void main( String[] args ) 
    {
        Counting arranjo = new Counting( 100 );
        Log log = new Log( );

        System.out.println( "\n//----- COUNTING SORT -----//");

        arranjo.preencher_decrescente();

        System.out.println("\nARRANJO PRE ORDENACAO:");
        arranjo.mostrar( );

        log.inicioTempo();
        arranjo.sort( log );
        log.fimTempo();

        System.out.println("\nARRANJO POS ORDENACAO:");
        arranjo.mostrar( );

        log.mostrar( );
    } // end main ( )    
}

File: Codigos/AEDs/Algoritmos de Ordenacao/C/Counting.c
#include "Geracao.h"
#include "Log.h"

int getMaior ( int* arranjo, int tamanho, Log* log ) 
{
    int maior = arranjo[0];
    for( int i = 0; i < tamanho; i = i + 1  ) 
    {
        incrementarComparacoes( log, 1 );
        if( maior < arranjo[i] ) 
        {
            maior = arranjo[i];
            incrementarMovimentacoes( log, 1 );
        } // end if
    } // end for
    return ( maior );
} // end getMaior ( )

void sort ( int* arranjo, int tamanho, Log* log )
{
    //Array para contar o numero de ocorrencias de cada elemento
    int tamCount = getMaior(arranjo, tamanho, log) + 1;

    int count[tamCount];
    int ordenado[tamanho];
    incrementarSubarranjos( log, 2 );

    //Inicializar cada posicao do arranjo de contagem 
    for( int i = 0; i < tamCount; 
        count[i] = 0, incrementarMovimentacoes(log, 1), i = i + 1 );

    //Agora, o count[i] contem o numero de elemento iguais a i
    for( int i = 0; i < tamanho; 
        count[arranjo[i]]++, i = i + 1 );

    //Agora, o count[i] contem o numero de elemento menores ou iguais a i
    for( int i = 1; i < tamCount; 
        count[i] += count[i-1], incrementarMovimentacoes(log, 1), i = i + 1 );

    //Ordenando
    for( int i = tamanho-1; i >= 0; 
        ordenado[count[arranjo[i]]-1] = arranjo[i], incrementarMovimentacoes(log, 1), count[arranjo[i]]--, i--);

    //Copiando para o arranjo original
    for( int i = 0; i < tamanho; 
        arranjo[i] = ordenado[i], incrementarMovimentacoes(log, 1), i = i + 1 );
} // end sort ( )

int main ( void )
{
    Geracao* arranjo = new_Geracao( 100 );
    Log*     log     = new_Log ( );

    printf( "\n//----- COUNTING SORT -----//\n");

    preencher_aleatorio ( arranjo, 200 );

    printf("\nARRANJO PRE ORDENACAO:\n");
    mostrar ( arranjo );

    inicioTempo ( log );
    sort ( arranjo->arranjo, arranjo->tamanho, log );
    fimTempo ( log );

    printf("\nARRANJO POS ORDENACAO:\n");
    mostrar ( arranjo );

    registrar ( log );

    delete_Log ( log );
    delete_Geracao ( arranjo );
} // end main ( )